### PR TITLE
Add "Search npm" extension

### PR DESCRIPTION
**Pull request by mrmartineau**
## Description

Extension to search for npm package using [npms.io](https://npms.io)

### Features

- search npm packages using [npms.io](https://npms.io)
- hit <kbd>ENTER</kbd> to view the package's homepage
- view the package's [bundlephobia.com](https://bundlephobia.com) cost. hit <kbd>⌘</kbd> <kbd>ENTER</kbd>
- view the package's README in Raycast
- view the package's Snyk vulnerability report
- view the package source code on GitHub.dev & CodeSandbox
- copy yarn/npm install commands to clipboard

## Type of change

- New extension

## Screencast

https://user-images.githubusercontent.com/64883/137721894-8860d0f6-73c9-4654-b232-96daae47235b.mp4

## Checklist

- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

### CODE REVIEW

File: extensions/search-npm/package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "search-npm",
+  "title": "Search npm",
```
Comment (thomaspaulmann) [inline extensions/search-npm/package.json:3]: The title is used in the store. I would suggest to use something like this.

```suggestion
  "title": "npm Package Search",
```

File: extensions/search-npm/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  "description": "",
```
Comment (thomaspaulmann) [inline extensions/search-npm/package.json:4]: Can you please provide a description? Again, this is used in the store.
```suggestion
  "description": "Search for npm package information from npms.io",
```

File: extensions/search-npm/package.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  "icon": "command-icon.png",
+  "author": "mrmartineau",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "index",
+      "title": "Search npm",
+      "subtitle": "Search for npm package information",
```
Comment (thomaspaulmann) [inline extensions/search-npm/package.json:12]: This fits better our UI guidelines: 
```suggestion
      "title": "Search Packages",
      "subtitle": "npm",
```

File: extensions/search-npm/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    >
+      {results?.length ? (
+        results.map((result) => {
+          if (!result.package.links.repository) {
+            return null
+          }
+          return <PackageListItem key={result.package.name} result={result} />
+        })
+      ) : (
+        <List.Item id="Empty" title="No results" icon={Icon.XmarkCircle} />
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:44]: We show a view if there aren't any results. You shouldn't need this.

File: extensions/search-npm/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      id={pkg.name}
+      key={pkg.name}
+      title={pkg.name}
+      subtitle={pkg.description}
+      icon={Icon.ArrowRight}
+      accessoryTitle={`v${pkg.version}`}
+      keywords={pkg.keywords}
+      actions={
+        <ActionPanel>
+          <OpenInBrowserAction url={pkg.links.repository} title="Repository" />
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:71]:
```suggestion
          <OpenInBrowserAction url={pkg.links.repository} title="Open Repository" />
```

File: extensions/search-npm/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+          <OpenInBrowserAction
+            url={`https://bundlephobia.com/package/${pkg.name}`}
+            title="Bundlephobia cost"
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:74]:
```suggestion
            title="Open Bundlephobia"
```

File: extensions/search-npm/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            icon={Icon.QuestionMark}
+          />
+          {isGithubRepo && parsedRepo?.owner && parsedRepo?.name ? (
+            <PushAction
+              title="View readme"
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:79]:
```suggestion
              title="Open README"
```

File: extensions/search-npm/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+              target={<Readme user={parsedRepo.owner} repo={parsedRepo.name} />}
+              icon={Icon.Eye}
+            />
+          ) : null}
+          {pkg.links.homepage !== pkg.links.repository ? (
+            <OpenInBrowserAction url={pkg.links.homepage} title="Homepage" />
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:85]:
```suggestion
            <OpenInBrowserAction url={pkg.links.homepage} title="Open Homepage" />
```

File: extensions/search-npm/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+          ) : null}
+          <OpenInBrowserAction url={pkg.links.npm} title="npm" />
+          <OpenInBrowserAction
+            url={`https://snyk.io/vuln/npm:${pkg.name}`}
+            title="Snyk vulnerability check"
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:90]: Maybe you can add a Snyk icon to this action? Same for the other bits. 

```suggestion
            title="Open Snyk Vulnerability Check"
```

File: extensions/search-npm/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+          />
+          <CopyToClipboardAction
+            title={`Copy "yarn install ${pkg.name}"`}
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:93]:
```suggestion
            title={`Copy Yarn Install Command`}
```

File: extensions/search-npm/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            content={`yarn install ${pkg.name}`}
+          />
+          <CopyToClipboardAction
+            title={`Copy "npm install ${pkg.name}"`}
```
Comment (thomaspaulmann) [inline extensions/search-npm/src/index.tsx:97]:
```suggestion
            title={`Copy npm Install Command`}
```

**Review by thomaspaulmann — CHANGES_REQUESTED** [review]
This is really cool! I can't wait to find my npm packages quicker with Raycast :-) I left a few inline comments to change the naming to follow our UI guidelines, which is most often title case. 

Can you please apply the suggestions and then we should be good to go.

File: extensions/search-npm/src/index.tsx (hunk 9 above)
Comment (mrmartineau) [reply extensions/search-npm/src/index.tsx:90]: I tried adding Snyk's icon and it is too complex to look good at this small size. Here's the best version that I could find:

image

Comment (FezVrasta) [reply extensions/search-npm/src/index.tsx:90]: They have a Press Kit with a monochrome version that may look good

https://snyk.io/press-kit/

Comment (mrmartineau) [reply extensions/search-npm/src/index.tsx]: Thanks, but I tried it and it still looked awful. I have committed the version in the screenshot above for now..

**Comment by mrmartineau** [comment]
@thomaspaulmann I updated the PR based on your comments. I also added a couple of extra features:

- view code in github.dev
- view code in CodeSandbox

**Review by thomaspaulmann — APPROVED** [review]
Thanks for applying all the changes and adding even more features. Let's get this into the store. If you have something new, just open another PR.

**Comment by PitNikola** [comment]
@mrmartineau, cool extension. Any idea why it doesn't find "@raycast/api" search string?

CleanShot 2021-10-20 at 10 14 29@2x

CleanShot 2021-10-20 at 10 15 09@2x

**Comment by mrmartineau** [comment]
@PitNikola 
> cool extension.
Thanks..

>  Any idea why it doesn't find "@raycast/api" search string?
The search is based on npms.io which also does not return any results for that query, see https://npms.io/search?q=%40raycast%2Fapi

```
❯ curl "https://api.npms.io/v2/search?q=@raycast/api"
{"total":0,"results":[]}⏎
```

Perhaps the special chars need to be escaped..

**Pull request merged**